Create video workflow documentation (#2010)
mitodl/ocw-studio: videos/README.md

# Video Workflow

This document describes the components of the video workflow for OCW.

**SECTIONS**

1. [Overview](#overview)
1. [Google Drive Sync and AWS Transcoding](#google-drive-sync-and-aws-transcoding)
1. [YouTube Submission](#youtube-submission)
1. [Captioning and 3Play Transcript Request](#captioning-and-3play-transcript-request)
1. [Completing the Workflow](#completing-the-workflow)
1. [Management Commands](#management-commands)

# Overview

> This assumes that AWS Media Convert, Google Drive sync, YouTube integration, and 3Play submission are all enabled, which is required for the video workflow.

Reviewer: Not really necessary, but might be nice to have links to the relevant sections in the other README here.

- Upload a video with the name `<video_name>.mp4` to the `videos_final` folder on Google Drive. If there are pre-existing captions that should be uploaded with the video (as opposed to requesting captions/transcript from 3Play), then these should be named _exactly_ `<video_name>_captions.vtt` and `<video_name>_transcript.pdf`, and uploaded into the `files_final` folder on Google Drive.
- Sync using the Studio UI. This uploads the video to S3.
- As soon as the upload to S3 is complete, Studio initiates a celery task to submit the video to the AWS Media Convert service.
- Once trancoding is complete, the video is uploaded to YouTube (set as unlisted prior to the course being published).
- After the video has been successfully uploaded to YouTube, and if there are no pre-existing captions, Studio sends a transcript request to 3Play.
- Once 3Play completes the transcript job, the captions (`.vtt` format) and transcript (`.pdf` format) are fetched and associated with the video.
- When the course is published, the video metadata and YouTube metadata are updated, and the YouTube video is set to public.

# Google Drive Sync and AWS Transcoding

> Users upload videos in `.mp4` format to the `videos_final` folder. Whether a file is located in this folder is used for defining the `is_video` property defined [here](https://github.com/mitodl/ocw-studio/blob/master/gdrive_sync/models.py).

Reviewer: I think other video formats are also accepted?

> The `TranscodeJobView` endpoint (defined [here](https://github.com/mitodl/ocw-studio/blob/master/videos/views.py)) listens for the webhook that the transcoding job is complete.

Reviewer: "listens for the webhook that is sent when the transcoding job is complete"

# YouTube Submission

Videos are uploaded to YouTube via the `resumable_upload` function, defined [here](https://github.com/mitodl/ocw-studio/blob/master/videos/youtube.py).

# Captioning and 3Play Transcript Request

If there are no pre-existing captions, a 3Play transcript request is generated. This is done via the `threeplay_transcript_api_request` function (defined [here](https://github.com/mitodl/ocw-studio/blob/master/videos/threeplay_api.py)).

# Completing the Workflow

> The webhooks tests are in [this folder](https://github.com/mitodl/ocw-studio/tree/master/test_videos_webhook).

Reviewer: Webhook tests seem like a different topic than workflow completion

# Video Workflow

This document describes the components of the video workflow for OCW.

**SECTIONS**

1. [Overview](#overview)
1. [Google Drive Sync and AWS Transcoding](#google-drive-sync-and-aws-transcoding)
1. [YouTube Submission](#youtube-submission)
1. [Captioning and 3Play Transcript Request](#captioning-and-3play-transcript-request)
1. [Completing the Workflow](#completing-the-workflow)

> 1. [Management Commands](#management-commands)

Reviewer 2: It's amazing how GitHub renders the correct bullet numbers. I did not know this was possible, I've been doing unnecessary manual labor. 🤣

- As soon as the upload to S3 is complete, Studio initiates a celery task to submit the video to the AWS Media Convert service.
- Once trancoding is complete, the video is uploaded to YouTube (set as unlisted prior to the course being published).
- After the video has been successfully uploaded to YouTube, and if there are no pre-existing captions, Studio sends a transcript request to 3Play.
- Once 3Play completes the transcript job, the captions (`.vtt` format) and transcript (`.pdf` format) are fetched and associated with the video.
- When the course is published, the video metadata and YouTube metadata are updated, and the YouTube video is set to public.

# Google Drive Sync and AWS Transcoding

Users upload videos in a valid video format to the `videos_final` folder. Whether a file is located in this folder is used for defining the `is_video` property defined [here](/gdrive_sync/models.py). The file is processed using the `process_drive_file` function [here](/gdrive_sync/tasks.py), which triggers the `stream_to_s3` and `transcode_gdrive_video` functions [here](/gdrive_sync/api.py), which submit the AWS MediaConvert transcoding job.

> The parameters of the AWS transcode request are defined through the AWS interface, and the role is defined [here](https://github.com/mitodl/ol-infrastructure/blob/main/src/ol_infrastructure/applications/ocw_studio/__main__.py). Some example JSONs used for triggering MediaConver job are in [this folder](/test_videos_webhook/).

Reviewer 3: ```suggestion
The parameters of the AWS transcode request are defined through the AWS interface, and the role is defined [here](https://github.com/mitodl/ol-infrastructure/blob/main/src/ol_infrastructure/applications/ocw_studio/__main__.py). Some example JSONs used for triggering MediaConvert job are in [this folder](/test_videos_webhook/).
```

The `TranscodeJobView` endpoint (defined [here](/videos/views.py)) listens for the webhook that is sent when the transcoding job is complete.

# YouTube Submission

> Videos are uploaded to YouTube via the `resumable_upload` function, defined [here](/videos/youtube.py).

Reviewer 2: ```suggestion
Videos are uploaded to YouTube via the `resumable_upload` function, defined [here](/videos/youtube.py). After upload, YouTube takes some time to process the videos. We use a periodic task `update_youtube_statuses`, defined [here](/videos/tasks.py), to check the completion of the process.
```

# Captioning and 3Play Transcript Request

> If there are no pre-existing captions, a 3Play transcript request is generated. This is done via the `threeplay_transcript_api_request` function (defined [here](videos/threeplay_api.py)).

Reviewer 2: ```suggestion
If there are no pre-existing captions, a 3Play transcript request is generated. This is done via the `threeplay_transcript_api_request` function (defined [here](videos/threeplay_api.py)). 3Play might take some time to generate the captions. We use a periodic task `update_transcripts_for_video`, defined [here](/videos/tasks) to check if the generation is complete.
```

# Completing the Workflow

> Once the workflow is completed, the `Video` and `WebsiteContent` objects get saved. The only remaining steps are triggered on course publish: updating the video metadata via `update_transcripts_for_website`, defined [here](/videos/tasks.py) and updating the YouTube metadata via `update_youtube_metadata`, defined [here](/videos/youtube.py).

Reviewer 2: We update related DB objects in every intermediate step. So it would be better to rephrase the first sentence to avoid confusion.

```suggestion
Once the workflow is completed, we have the updated `Video` and `WebsiteContent` objects. The only remaining steps are triggered on course publish: updating the video metadata via `update_transcripts_for_website`, defined [here](/videos/tasks.py), and updating the YouTube metadata via `update_youtube_metadata`, defined [here](/videos/youtube.py).
```

- On any publish action, the video metadata and YouTube metadata are updated, assuming the information has been received from the external services.
- The YouTube video is set to public once the course has been published to live/production.

# Google Drive Sync and AWS Transcoding

Users upload videos in a valid video format to the `videos_final` folder. Whether a file is located in this folder is used for defining the [is_video property](/gdrive_sync/models.py). The file is processed using the [process_drive_file function](/gdrive_sync/tasks.py), which triggers the [`stream_to_s3` and `transcode_gdrive_video` functions](/gdrive_sync/api.py), which submit the AWS MediaConvert transcoding job.

The parameters of the AWS transcode request are defined through the AWS interface, and the role is defined [here](https://github.com/mitodl/ol-infrastructure/blob/main/src/ol_infrastructure/applications/ocw_studio/__main__.py). Some example JSONs used for triggering MediaConvert job are in [this folder](/test_videos_webhook/).

The [`TranscodeJobView` endpoint](/videos/views.py) listens for the webhook that is sent when the transcoding job is complete.

# YouTube Submission

> Videos are uploaded to YouTube via the [`resumable_upload` function](/videos/youtube.py). The [YouTube upload success notification](/videos/templates/mail/youtube_upload_success/body.html) is sent by email when the [`update_youtube_statuses`](/videos/tasks.py) task is complete; exceptions in this task trigger the [YouTube upload failure notification](/videos/templates/mail/youtube_upload_failure/body.html). When the course is published to draft/staging, the video is set to private. However, when it is published to live/production, the video is made public on YouTube, via the [`update_youtube_metadata` function](/videos/youtube.py).

Reviewer 3: Consider:

When a video is made public on YouTube, all YouTube subscribers will be notified. There are nearly 5 million subscribers to the OCW YouTube channel, so be careful with this setting.

# Captioning and 3Play Transcript Request

If there are no pre-existing captions, a 3Play transcript request is generated. This is done via the [`threeplay_transcript_api_request` function](videos/threeplay_api.py).

# Completing the Workflow

Once the workflow is completed, the updates to the `Video` and `WebsiteContent` objects are nearly complete. The only remaining steps are triggered on course publish: updating the video metadata via [`update_transcripts_for_website`](/videos/tasks.py) and updating the YouTube metadata via [`update_youtube_metadata`](/videos/youtube.py).

# Management Commands

> In cases where something may have gone wrong with the data, often due to legacy data issues, there are management commands that can be run to resolve them. The commands are defined [here](/videos/management/commands/).

Reviewer 4: Maybe it would be best to briefly go over some common failure points and how these management commands can address them?
